js/src/forum/states/DiscussionListState.ts

```diff
@@ -0,0 +1,111 @@
+import PaginatedListState from '../../common/states/PaginatedListState';
+import Discussion from '../../common/models/Discussion';
+
+export default class DiscussionListState extends PaginatedListState<Discussion> {
+  constructor(params: any, page: number) {
+    super(20);
+
+    this.params = params;
+
+    this.location = { page };
+  }
+
+  get type(): string {
+    return 'discussions';
+  }
+
+  requestParams() {
+    const params: any = { include: ['user', 'lastPostedUser'], filter: {} };
+
+    params.sort = this.sortMap()[this.params.sort];
+
+    if (this.params.q) {
+      params.filter.q = this.params.q;
+
+      params.include.push('mostRelevantPost', 'mostRelevantPost.user');
+    }
+    return params;
+  }
+
+  protected loadPage(page: number = 1): any {
+    const preloadedDiscussions = app.preloadedApiDocument();
+
+    if (preloadedDiscussions) {
+      this.initialLoading = false;
+
+      return Promise.resolve(preloadedDiscussions);
+    }
+
+    return super.loadPage(page);
+  }
+
+  /**
+   * Get a map of sort keys (which appear in the URL, and are used for
+   * translation) to the API sort value that they represent.
+   */
+  sortMap() {
+    const map: any = {};
+
+    if (this.params.q) {
+      map.relevance = '';
+    }
+    map.latest = '-lastPostedAt';
+    map.top = '-commentCount';
+    map.newest = '-createdAt';
+    map.oldest = 'createdAt';
+
+    return map;
+  }
+
+  /**
+   * In the last request, has the user searched for a discussion?
+   */
+  isSearchResults() {
+    return !!this.params.q;
+  }
+
+  removeDiscussion(discussion: Discussion) {
```

Comment:
Do we want to include the add/remove methods in the main class? I think this is a pattern we are likely to see relatively often.

Reply:
Hm, we might want a separate abstract class that implements them instead. I don't like having so many methods in the main class tbh.

Reply:
Fair enough. We can tackle this one when it's actually needed.